components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/autofill/LoginStorageDelegate.kt

```diff
@@ -0,0 +1,121 @@
+import mozilla.components.service.sync.logins.ServerPassword
+import org.mozilla.geckoview.GeckoResult
+
+// Temporary Interface before lands in GV
+internal interface LoginDelegate {
+    fun onLoginUsed(login: Login)
+    fun onFetchLogins(domain: String): GeckoResult<Array<Login>>
+    fun onLoginSave(login: Login)
+}
+
+/**
+ * [LoginDelegate /* TODO UPDATE */] implementation.
+ *
+ * When GV needs to update or retrieve information about stored logins, it will call into this
+ * class.
+ *
+ * App will have to instantiate this and set it on the runtime and pass in the [LoginsStorage] and
+ * [SecureAbove22Preferences] with a key that conforms to "passwords".  This lets GV delegate to
+ * the client for storage.
+ */
+class LoginStorageDelegate(
+    private val loginStorage: LoginsStorage,
+    private val keyStore: SecureAbove22Preferences
+) : LoginDelegate {
+    override fun onLoginUsed(login: Login) {
+        val passwordsKey = keyStore.getString(PASSWORDS_KEY)
+        val guid = login.guid
+        // TODO should this short on an empty string?
+        if (passwordsKey == null || guid == null || guid.isEmpty()) return
+        with(loginStorage) {
+            ensureUnlocked(passwordsKey)
+            touch(guid)
+            lock()
+        }
+    }
+
+    override fun onFetchLogins(domain: String): GeckoResult<Array<Login>> {
+        val passwordsKey =
+            keyStore.getString(PASSWORDS_KEY) ?: return GeckoResult.fromValue(arrayOf())
+        loginStorage.ensureUnlocked(passwordsKey).also {
+            val result = GeckoResult.fromValue(loginStorage.getByHostname(domain).map { // TODO getByHostname -> getByBaseDomain
+                Login(
+                    guid = it.id,
+                    origin = it.hostname,
+                    formActionOrigin = it.formSubmitURL,
+                    httpRealm = it.httpRealm,
+                    username = it.username ?: "", // TODO this should be nonnull after an incoming AS update
+                    password = it.password
+                )
+            }.toTypedArray())
+            loginStorage.lock()
+            return result
+        }
+    }
+
+    // TODO double check that we're locking correctly.  Also, does this need to be synchronized?
+    // Request to save or update the given login.
+    @Synchronized
+    override fun onLoginSave(login: Login) {
+        val passwordsKey = keyStore.getString(PASSWORDS_KEY) ?: return // TODO should we fail fast if this is null?
+        loginStorage.ensureUnlocked(passwordsKey)
+        val guid = login.guid
+        val serverPassword = guid?.let { loginStorage.get(it) }
+
+        if (guid != null && serverPassword != null) {
+            loginStorage.update(serverPassword.mergeWithLogin(login))
+        } else {
+            loginStorage.add(
+                ServerPassword(
+                    id = "", // TODO ask if this is right? Pass empty string if we don't have it?  If so, ask for it to be null
```

Comment:
@vladikoff does passing an empty string for a new entry make sense?

Reply:
Hm looking at the tests, I don't think so. @thomcc confirm? https://github.com/mozilla/application-services/blob/e81ff0d715758da8ed5e741c2fea73faa3a74e18/components/logins/android/src/main/java/mozilla/appservices/logins/ServerPassword.kt#L18

Reply:
@grigoryk believes that passing an empty string here will cause a guid to be generated.  He suggests that we try that for now.  I'll test that and update here.

Reply:
Yeah, i think passing an empty string will get the rust code to generate a guid.

After our chat with @mnoorenberghe , i think this is definitely more nuanced.

You have four cases, I think:
- update an existing record with a guid
- save a new record (no guid)
- save a new record (has a guid, but user changed the username, so it's a new login now)
- update an existing record with a guid, which had an empty username, and user added a username